### PR TITLE
Experimental flag to attach row count metadata as part of `dagster-dbt` execution

### PR DESCRIPTION
**Pull request by benpankow**
## Summary

Adds a new `fetch_table_metadata` experimental flag to `DbtCliResource.cli` which allows us to fetch `dagster/total_row_count` (introduced in #21524) to dbt-built tables:

```python
@dbt_assets(manifest=dbt_manifest)
def jaffle_shop_dbt_assets(
    context: AssetExecutionContext,
    dbt: DbtCliResource,
):
    yield from dbt.cli(
        ["build"],
        context=context,
        fetch_table_metadata=True,
    ).stream()
```

Screenshot 2024-05-03 at 11 03 19 AM

Under the hood, this PR uses dbt's `dbt.adapters.base.impl.BaseAdapter` abstraction to let Dagster connect to the user's warehouse using the dbt-provided credentials. Right now, we just run a simple `select count(*)` on the tables specified in each `AssetMaterialization` and `Output`, but this lays some groundwork we could use for fetching other data as well.

There are a few caveats:
- When using duckdb, we wait for the dbt run to conclude, since duckdb does not allow simultaneous connections when a write connection is open (e.g. when dbt is running)
- We don't query row counts on views, since they may include non-trivial sql which could be expensive to query

## Test Plan

Tested locally w/ duckdb, bigquery, and snowflake. Introduced basic pytest test to test against duckdb.

### CODE REVIEW

**Comment by benpankow** [comment]
* **#21542** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21542?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @benpankow and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 1]
```diff
@@ -505,6 +511,68 @@ def _build_column_lineage_metadata(
             )
 
 
+DbtDagsterEventType = Union[Output, AssetMaterialization, AssetCheckResult, AssetObservation]
+
+
+class DbtPostprocessingAsyncTask(ABC):
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:517]: I don't think we would want to generalize to subclass a stateless `AsyncTask`. 

In the future, for efficiency, we would actually want this to be stateful (e.g. to store adapter connections and cached information from API calls).

Let's remove this for now and just have a single function that does post-processing on an event.

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 2]
```diff
@@ -1116,6 +1291,13 @@ def _initialize_adapter(self, args: Sequence[str]) -> BaseAdapter:
         project = load_project(self.project_dir, False, profile, {})
         config = RuntimeConfig.from_parts(project, profile, flags)
 
+        from dbt.adapters.duckdb.credentials import DuckDBCredentials
+
+        if isinstance(config.credentials, DuckDBCredentials):
+            if not config.credentials.config_options:
+                config.credentials.config_options = {}
+            config.credentials.config_options["access_mode"] = "READ_ONLY"
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:1299]: Could you explain what happening here?

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/utils.py [hunk 3]
```diff
@@ -282,3 +283,20 @@ def parse_manifest(path: str, target_path: str = DEFAULT_DBT_TARGET_PATH) -> Map
             return json.load(file)
     except FileNotFoundError:
         raise DagsterDbtCliOutputsNotFoundError(path=manifest_path)
+
+
+def get_future_completion_state_or_err(futures: List[Union[Future, Any]]) -> bool:
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/utils.py:288]: Could we instead use https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.wait with some timeout for similar functionality?

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 4]
```diff
@@ -669,13 +763,94 @@ def stream(
...
-                dagster_dbt_translator=self.dagster_dbt_translator,
-                context=self.context,
-                target_path=self.target_path,
-            )
+        # We keep a list of emitted Dagster events and pending futures which augment
+        # emitted events with additional metadata. This ensures we can yield events in the order
+        # they are emitted by dbt.
+        output_events_and_futures: List[Union[Future, DbtDagsterEventType]] = []
+
+        with pushd(str(self.project_dir)):
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:771]: Could you explain the `pushd` here?

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 5]
```diff
@@ -669,13 +763,94 @@ def stream(
...
+            with ThreadPoolExecutor(max_workers=self.stream_events_threadpool_size) as executor:
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:772]: nit: you can combine contextmanagers in a single with block so we don't have excessive nesting.

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 6]
```diff
@@ -669,13 +763,94 @@ def stream(
...
+
+                def run_postprocessing_tasks(
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:774]: Let's separate out these functions as private methods to make this more legible

**Review by rexledesma — DISMISSED** [review]
Main thing here is that I'm not sold on this PostprocessingTask abstraction. Let's just define a single stateful transformation function.

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 7]
```diff
@@ -543,6 +613,7 @@ def run(
         raise_on_error: bool,
         context: Optional[OpExecutionContext],
         adapter: Optional[BaseAdapter],
+        postprocessing_tasks: Optional[List[DbtPostprocessingAsyncTask]],
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:616]: Just so we have maximum flexibility, we shouldn't expose this API to the user yet. Once we have confidence in this scheme working for out-of-the-box metadata (e.g. row count), then we can consider exposing it to the user so they also have more flexibility.

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py (hunk 2 above)
Comment (benpankow) [reply python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:1299]: attempts to get duckdb adapter to play nice - concurrently, duckdb only accepts either:
a) one read-write connection
b) many read-only connections

this lets us parallelize our accesses for postprocessing in the duckdb case - sadly don't think we can run any of these tasks simultaneously to the dbt build since that's read-write

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py (hunk 4 above)
Comment (benpankow) [reply python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:771]: needed to get the duckdb adapter to work correctly - it tries to access the db file assuming the working directory is the project dir - seems like the DBT cli does this in its process, but we need to do the same if we want to use the adapter ourselves

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py (hunk 1 above)
Comment (benpankow) [reply python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py]: Yeah, I agree it's probably premature, even if we want to write a few different tasks ourselves I think we don't need the abstraction

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/utils.py (hunk 3 above)
Comment (benpankow) [reply python_modules/libraries/dagster-dbt/dagster_dbt/core/utils.py:288]: Probably, but I think we'd still have to do some custom logic around bookkeeping on event log/Future order. Right now the waiting/timeout is handled in the main loop in `stream` since we just want to wait on the next incomplete event log entry's `Future` (so we can yield it to user code). This way we don't have to worry about yielding things in the wrong order.

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 8]
```diff
@@ -543,6 +554,7 @@ def run(
         raise_on_error: bool,
         context: Optional[OpExecutionContext],
         adapter: Optional[BaseAdapter],
+        fetch_table_metadata: bool,
```
Comment (sryza) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:557]: A little worried about booleans like this getting out of control. Not entirely sure if this is better, but would something like the following be feasible:

```python
@dbt_assets
def my_dbt_assets(dbt):
    yield from dbt.run(...).fetch_row_counts()
```

More verbose but also more composable. And would potentially offer the ability to add more control over what subset of models we fetch row counts for, if fetching them all is too expensive.

Comment (rexledesma) [reply python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:557]: I agree. I don't particularly like the boolean pattern myself.

Since `DbtCliInvocation` is the intermediate object, it makes sense to let it be the object that dictates how the default translation from dbt events to Dagster events happens.

Just making this a bit more explicit: `.stream()` should still be called, but only when `DbtCliInvocation` is finalized. `.fetch_XXX` are methods on `DbtCliInvocation` that return `Self`.

```python
@dbt_assets(...)
def my_dbt_assets(context: AssetExecutionContext, dbt: DbtCliResource):
    yield from (
        dbt.run(...)
        .fetch_column_schema() # Fetch column schema in Python, rather than in dbt jinja 
        .fetch_row_counts()    # This PR
        .stream()
    )

File: python_modules/dagster/dagster/_core/definitions/events.py [hunk 9]
```diff
@@ -149,6 +149,17 @@ def __eq__(self, other: object) -> bool:
             and self.tags == other.tags
         )
 
+    def with_metadata(self, metadata: Optional[Mapping[str, RawMetadataValue]]) -> "Output":
```
Comment (rexledesma) [inline python_modules/dagster/dagster/_core/definitions/events.py:152]: We should separate this out into its own PR, since this is a core lib change.

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 10]
```diff
@@ -633,6 +646,98 @@ def is_successful(self) -> bool:
...
+        adapter = check.not_none(self.adapter)
+
+        dbt_resource_props = self._get_dbt_resource_props_from_event(event)
+        is_view = dbt_resource_props["config"]["materialized"] == "view"
+
+        # Avoid counting rows for views, since they may include complex SQL queries
+        # that are costly to execute. We can revisit this in the future if there is
+        # a demand for it.
+        if is_view:
+            return event
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:684]: This should be communicated to the user somehow, through a single initial logging statement.

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 11]
```diff
@@ -633,6 +646,98 @@ def is_successful(self) -> bool:
...
+
+        # If the adapter is DuckDB, we need to wait for the dbt CLI process to complete
+        # so that the DuckDB lock is released
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:687]: Let's add some more comments about why this is an edge case for only DuckDB.

Other adapters should be fine in supporting multiple connections.

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 12]
```diff
@@ -633,6 +646,98 @@ def is_successful(self) -> bool:
...
+
+            if isinstance(adapter, DuckDBAdapter):
+                self._dbt_run_thread.result()
+        except ImportError:
+            pass
+
+        unique_id = dbt_resource_props["unique_id"]
+        table_str = f"{dbt_resource_props['database']}.{dbt_resource_props['schema']}.{dbt_resource_props['name']}"
+
+        adapter = check.not_none(self.adapter)
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:699]: this already occurs at the top
```suggestion
```

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 13]
```diff
@@ -633,6 +646,98 @@ def is_successful(self) -> bool:
...
+            query_result = adapter.execute(
+                f"""
+                    SELECT
+                    count(*)
+                    FROM
+                    {table_str}
+                """,
+                fetch=True,
+            )
+        row_count = query_result[1][0][0]
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:711]: nit: can we alias `count(*)` so that we don't have to do this array indexing? Or could we use result de-structuring instead?

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 14]
```diff
@@ -633,6 +646,98 @@ def is_successful(self) -> bool:
...
+        additional_metadata = {**TableMetadataSet(total_row_count=row_count)}
+
+        if isinstance(event, Output):
+            return event.with_metadata(metadata={**event.metadata, **additional_metadata})
+        else:
+            return event._replace(metadata={**event.metadata, **additional_metadata})
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:717]: This signals that the core Dagster `XXXEvent` classes should have a more unified inheritance pattern.

We should file an issue and mark this as one code path to consolidate.

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 15]
```diff
@@ -669,14 +774,47 @@ def stream(
...
+            # once they are ready to be emitted
+            event_to_emit_idx = 0
+            while True:
+                all_work_complete = get_future_completion_state_or_err(
+                    [self._dbt_run_thread, *output_events_and_futures]
+                )
+                if all_work_complete and event_to_emit_idx >= len(output_events_and_futures):
+                    break
+
+                if event_to_emit_idx < len(output_events_and_futures):
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:802]: nit: invert conditional logic, there's a lot of nesting here

**Review by rexledesma — CHANGES_REQUESTED** [review]
Main comments:
- Divest from the boolean pattern, to more of a composable builder pattern
- The `with_metadata` change should be a separate PR. There should be systematic refactoring of our event inheritance to ensure metadata can be overridden in an ergonomic way.

File: python_modules/dagster/dagster/_core/definitions/events.py (hunk 9 above)
Comment (benpankow) [reply python_modules/dagster/dagster/_core/definitions/events.py]: #21666

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py (hunk 10 above)
Comment (benpankow) [reply python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:683]: It's mentioned in the `fetch_row_counts` apidoc, do think that it also needs to be output at runtime?

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py (hunk 14 above)
Comment (benpankow) [reply python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:716]: #21667

**Comment by benpankow** [comment]
Updated to builder pattern

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py (hunk 10 above)
Comment (rexledesma) [reply python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:683]: Yeah, explicit communication is nice for observability

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 16]
```diff
@@ -633,6 +645,118 @@ def is_successful(self) -> bool:
...
+
+    @experimental
+    def fetch_row_counts(
+        self,
+    ) -> "DbtCliInvocation":
+        """Experimental functionality which will fetch row counts for materialized dbt
+        models in a dbt run once they are built. Note that row counts will not be fetched
+        for views, since this requires running the view's SQL query which may be costly.
+        """
+        return DbtCliInvocation(
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:748]: Can use [`dataclasses.replace`](https://docs.python.org/3/library/dataclasses.html#dataclasses.replace)

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 17]
```diff
@@ -633,6 +645,118 @@ def is_successful(self) -> bool:
...
+                    executor.submit(
+                        self._attach_post_materialization_metadata,
+                        event,
+                    )
+                )
+            else:
+                output_events_and_futures.append(event)
+
+    @experimental
+    def fetch_row_counts(
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:741]: Maybe remove the plural?

Also, I wonder if this method is ill-named.

The current method name insinuates that the fetch is happening after the method is called, which is not true. It is setting an attribute for `.stream()` to read from.

Some suggestions:

- `with_row_count`
- `with_row_count_retrieval`
- `with_row_count_fetch`
- `enable_row_count_retrieval`
- `enable_fetch_row_count`

Comment (benpankow) [reply python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:741]: Yeah, I like `enable_fetch_row_count`

Comment (sryza) [reply python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:741]: Maybe worth chatting about in standup, but the plural sounds right to me. Multiple row counts are being fetched.

Comment (sryza) [reply python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py]: > insinuates that the fetch is happening after the method is called, which is not true.

Should we definitely not do the fetch when the method is called?

Comment (benpankow) [reply python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py]: To do so we'd need to stream/process the dbt event stream in some form, which we typically don't do until the user calls `stream()`.

File: python_modules/libraries/dagster-dbt/dagster_dbt_tests/core/dbt_packages/test_row_count_postprocessing.py
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt_tests/core/dbt_packages/test_row_count_postprocessing.py:1]: I don't think this should be in the `dbt_packages/` directory since this functionality doesn't rely on the `dagster` dbt package.

**Review by rexledesma — APPROVED** [review]
I assume docs + API docstring examples are coming in a follow up PR
